How to ingest postgres queries into watchdog (#1151)
medic/cht-docs: content/en/apps/guides/hosting/monitoring/postgres-ingest.md

2. [Add new Postgres Exporter]({{< relref "#add-new-postgres-exporter" >}})
3. [Adding new scrape config]({{< relref "#adding-new-scrape-config" >}})
4. [Load new Compose files with existing ones]({{< relref "#load-new-compose-files-with-existing-ones" >}})
5. [Configure the dashboard]({{< relref "#configure-the-dashboard" >}})
6. [Save the dashboard for easy deployment and future updates]({{< relref "#save-the-dashboard-for-easy-deployment-and-future-updates" >}})

### Prepare query in config file

Add a YAML file for with your query called `~/custom-sql-queries.yml`. In this example we'll be using a query from the [App Monitoring Data Ingestion repo](https://github.com/medic/cht-app-monitoring-data-ingest/), but it can be any query as long as the user you're using in the next step has access to the database and table:

```yaml
dwh_impact_replication_failure:
  query: |
    SELECT 
      metric as reason,
      count as total
    FROM 
      public.app_monitoring_replication_failure_reasons
    WHERE 
      partner_name IN ('partner_name_here')
  metrics:
    - reason:
        usage: "LABEL"
        description: "Name of the failure"
    - total:

>         usage: "LABEL"

Author: Ah - right! Sorry for my confusion.

After my work on [the data ingest repo](https://github.com/medic/cht-app-monitoring-data-ingest/pull/104), I think `GAUGE` is correct.  Updated in af596909

```

This configuration will generate a metric named `dwh_impact_replication_failure_total` with a label named `reason` which contains the string key value identifying the aggregated reason for the given replication failures.  These metric/label names are fully customizable, but to avoid confusion you should follow the Prometheus [best practices](https://prometheus.io/docs/practices/naming/) when choosing names. 

### Add new Postgres Exporter

In a new file, `~/docker-compose.custom-sql.yml`, define your new Postgres exporter as well as add a mount to the existing Grafana and Prometheus services. Note that you will need to add the following environment variables to your Watchdog `~/cht-watchdog/.env` file:

* `CUSTOM_SQL_USER` - Postgres user to use when logging in
* `CUSTOM_SQL_PASS` - Password for `CUSTOM_SQL_USER` above
* `CUSTOM_SQL_SERVER` - URL or IP for your Postgres server
* `CUSTOM_SQL_PORT` - Port of server, defaults to `5432` it not declared.
* `CUSTOM_SQL_DATABASE` - Actual string of database name (eg `extra_monitoring` or `health_stats`), will be different for each install.

```yaml
services:

  prometheus:
    volumes:

>       - ./scrape_config.custom-sql.yml:/etc/prometheus/scrape_configs/custom-sql.yml:ro

Author: yes- good call!

    command:
      # disables the collection of all metrics except for custom queries 
      - '--no-collector.database'
      - '--no-collector.postmaster'
      - '--no-collector.process_idle'
      - '--no-collector.replication'
      - '--no-collector.replication_slot'
      - '--no-collector.stat_bgwriter'
      - '--no-collector.stat_database'
      - '--no-collector.statio_user_tables'
      - '--no-collector.stat_statements'
      - '--no-collector.stat_user_tables'
      - '--disable-default-metrics'
      - '--disable-settings-metrics'
    volumes:
      - ./custom-sql-queries.yml:/custom-sql-queries.yml
    environment:
      DATA_SOURCE_NAME: "postgresql://${CUSTOM_SQL_USER:-NO DB USER SPECIFIED}:${CUSTOM_SQL_PASS:-NO DB PASSWORD SPECIFIED}@${CUSTOM_SQL_SERVER:-.NO DB SERVER SPECIFIED}:${CUSTOM_SQL_PORT:-5432}/${CUSTOM_SQL_DATABASE:-.NO DB SPECIFIED}?sslmode=disable"
      PG_EXPORTER_EXTEND_QUERY_PATH: "/custom-sql-queries.yml"
    restart: always
    networks:
     - cht-watchdog-net
```

If there's a missing mount target, then Docker will error.  Avoid the error by creating an empty file called `~/extra-sql-dashboard.json`.  We'll use this in the last step below

> ### Adding new scrape config

Author: ah - right right! I have the justification earlier in the PR before I saw your good suggestion to remove the JSON mount.  I've removed this section  aa4c0ceb.

> ### Adding new scrape config

Author: and the above comment was for [this question](https://github.com/medic/cht-docs/pull/1151#discussion_r1313148797) - sorry! 

for your suggestion to rearrange - agreed! done in 0c2dce4b

1. [Prepare query in config file]({{< relref "#prepare-query-in-config-file" >}})
2. [Add new Postgres Exporter]({{< relref "#add-new-postgres-exporter" >}})
3. [Adding new scrape config]({{< relref "#adding-new-scrape-config" >}})
4. [Load new Compose files with existing ones]({{< relref "#load-new-compose-files-with-existing-ones" >}})
5. [Configure the dashboard]({{< relref "#configure-the-dashboard" >}})
6. [Save the dashboard for easy deployment and future updates]({{< relref "#save-the-dashboard-for-easy-deployment-and-future-updates" >}})

### Prepare query in config file

Add a YAML file for with your query called `~/extra-sql-queries.yml`. In this example we'll be using a query from the [App Monitoring Data Ingestion repo](https://github.com/medic/cht-app-monitoring-data-ingest/), but it can be any query as long as the user you're using in the next step has access to the database and table:

```yaml
replication_failure_reasons:
  query: |
    SELECT 
      metric as failure_type,
      sum(count) AS "count"
    FROM 
      public.app_monitoring_replication_failure_reasons
    WHERE 
      partner_name IN ('partner_name_here')
    GROUP BY 
      failure_type
    ORDER BY 

>       count DESC

Author: ah - gotcha  - you're right! This was an artifact of the more complex query with the `monitoring_urls` table which is not used here. nice catch!

  metrics:
    - failure:
        usage: "LABEL"
        description: "Name of the failure"

>     - sequence:

Author: nope - good call! I largely copied and pasted this from another example (watchdog core i'd guess).  poor research on my part!

>         usage: "TABLE"

Author: ah - gotcha!  i think I had conflated how this works when I was prototyping.  deferring to `LABEL` (and fixing a bad quote!) in 1d770f8 

thanks!

        description: "Replication failure reasons"
```

Note that you can use any value on the first line, seen as  `replication_failure_reasons` above.  It will be the field name we'll use in the 4th step below when exploring the new data in Grafana.

### Add new Postgres Exporter

In a new file  `~/extra-sql-compose.yml`, define your new Postgres exporter as well as add a mount to the existing Grafana and Prometheus services. Note that the `DATA_SOURCE_NAME` value will need to have the following variables added to you `.env` file you created to deploy Watchdog:

* `EXTRA_SQL_USER` - Postgres user to use when logging in
* `EXTRA_SQL_PASS` - Password for `EXTRA_SQL_USER` above
* `EXTRA_SQL_SERVER` - URL or IP for your Postgres server
* `EXTRA_SQL_PORT` - Port of server, defaults to `5432` it not declared.
* `EXTRA_SQL_DATABASE` - Actual string of database name (eg `extra_monitoring` or `health_stats`), will be different for each install.

```yaml
services:

  prometheus:
    volumes:

>       - ./extra-sql-prometheus.yml:/etc/prometheus/scrape_configs/extra-sql-prometheus.yml:ro

Reviewer: ```suggestion
      - ./scrape_config.custom-sql.yml:/etc/prometheus/scrape_configs/custom-sql.yml:ro
```

  grafana:
    volumes:

>       - ./extra-sql-dashboard.json:/etc/grafana/provisioning/dashboards/CHT/cht_admin_extra_sql.json:ro

Reviewer: ```suggestion
      - ./dashboard.custom-sql.json:/etc/grafana/provisioning/dashboards/CUSTOM/custom-sql.json:ro
```

I think we should put the dashboard in a separate directory than `CHT`. The idea is that everything (alerts, dashboards, etc) in the `CHT` directory come from Watchdog. Partners wanting to add additional custom dashboards, etc should put these in different directories.

Reviewer: Upon further thought, we might need to ignore this comment at least for now.  Adding a new directory to Grafana will require adding a new [`provider` instance](https://github.com/medic/cht-watchdog/blob/main/grafana/provisioning/dashboards/cht.yml).  I _think_ we could do that as needed by mapping in another `yml` config file to `/etc/grafana/provisioning/dashboards`, but I have not tried yet...

Author: ah - gotcha!  I'll leave as is, but just update file names per you suggestion - done in e90f21fa

>   extra_sql_exporter:

Reviewer: ```suggestion
  custom_sql_exporter:
```

    image: prometheuscommunity/postgres-exporter:latest
    command:
      # disables the collection of all metrics except for custom queries 
      - '--no-collector.database'
      - '--no-collector.postmaster'
      - '--no-collector.process_idle'
      - '--no-collector.replication'
      - '--no-collector.replication_slot'
      - '--no-collector.stat_bgwriter'
      - '--no-collector.stat_database'
      - '--no-collector.statio_user_tables'
      - '--no-collector.stat_statements'
      - '--no-collector.stat_user_tables'
      - '--disable-default-metrics'
      - '--disable-settings-metrics'
    volumes:

>       - ./extra-sql-queries.yml:/extra-sql-queries.yml

Reviewer: ```suggestion
      - ./custom-sql-queries.yml:/custom-sql-queries.yml
```

    environment:

>       DATA_SOURCE_NAME: "postgresql://${EXTRA_SQL_USER:-NO DB USER SPECIFIED}:${EXTRA_SQL_PASS:-NO DB PASSWORD SPECIFIED}@${EXTRA_SQL_SERVER:-.NO DB SERVER SPECIFIED}:${EXTRA_SQL_PORT:-5432}/${EXTRA_SQL_DATABASE:-.NO DB SPECIFIED}?sslmode=disable"

Reviewer: ```suggestion
      DATA_SOURCE_NAME: "postgresql://${CUSTOM_SQL_USER:-NO DB USER SPECIFIED}:${CUSTOM_SQL_PASS:-NO DB PASSWORD SPECIFIED}@${CUSTOM_SQL_SERVER:-.NO DB SERVER SPECIFIED}:${CUSTOM_SQL_PORT:-5432}/${CUSTOM_SQL_DATABASE:-.NO DB SPECIFIED}?sslmode=disable"
```

>       PG_EXPORTER_EXTEND_QUERY_PATH: "/extra-sql-queries.yml"

Reviewer: ```suggestion
      PG_EXPORTER_EXTEND_QUERY_PATH: "/custom-sql-queries.yml"
```

    restart: always
    networks:
     - cht-watchdog-net
```

> If there's a missing mount target, then Docker will error.  Avoid the error by creating an empty file called `~/extra-sql-dashboard.json`.  We'll use this in the last step below

Reviewer: ```suggestion
```

Is there a reason we are defining the docker-compose config in the instructions before setting up the scrape config or the dashboard?  To me, it is more confusing to be trying to map "future" files in the compose.yml vs just creating each of the config files up-front and then defining the docker-compose file...

Author: Yeah - it is a bit confusing, agreed. 

My thinking was that I wanted one edit to the compose file.  Once you had that in place, you could actually prototype the query, create the dashboard and save the JSON out to paste in the file. Seemed clean-ish.

Otherwise you have to:
1. set up the scrape and the query file, do most of the compose.
2. prototype the query, create the dashboard and save the JSON
3. create the JSON file
4. update the compose file to have JSON file mounted in grafana
5. restart and test it worked

so I'm shaving off a few steps as they felt dirty-ish.

I'll lightly push back but if you still disagree, I'll add more steps to the docs!

### Adding new scrape config

> Create the `~/extra-sql-prometheus.yml` file and point the config to our new Postgres Exporter which is `extra_sql_exporter:9187` on the last line below.  This will tell Prometheus to scrape the new data every 1 minute:

Reviewer: ```suggestion
Create the `~/scrape_config.custom-sql.yml` file and point the config to our new Postgres Exporter (`custom_sql_exporter:9187`).  This will tell Prometheus to scrape the new data every 1 minute:
```

```yaml
scrape_configs:

>   - job_name: 'extra-sql'

Reviewer: ```suggestion
  - job_name: 'custom-sql'
```

    scrape_interval: 1m
    static_configs:

>       - targets: ['extra_sql_exporter:9187']

Reviewer: ```suggestion
      - targets: ['custom_sql_exporter:9187']
```

```

> ### Load new Compose files with existing ones

Reviewer: ```suggestion
Launch Watchdog with the new compose file
```

Now that you've added the new configuration files, we can load it alongside the existing ones.  Assuming you've followed the [Watchdog Setup]({{< relref "apps/guides/hosting/monitoring/setup" >}}), this would be:

```shell
cd ~/cht-monitoring

> docker compose -f docker-compose.yml -f ../cadvisor-compose.yml -f ../extra-sql-compose.yml up -d

Reviewer: ```suggestion
docker compose -f docker-compose.yml -f ../docker-compose.custom-sql.yml up -d
```

```

### Configure the dashboard

Now that the new Postgres Exporter is running on your Watchdog instance and CHT Watchdog's Prometheus has additional scrape configs to ingest the new metrics, we can now visualize it in a Grafana Dashboard and then alert on it:

1. In the "Metric" field enter `replication_failure_reasons_count` from the step above where we defined `extra-sql-queries.yml`
2. Click the blue "Run query" in the upper right. 
3. We'll make this a table, but you can configure the dashboard as desired. 
4. Click "Add to dashboard"

![Grafana showing data data explorer](explore.png)

> ### Save the dashboard for easy deployment and future updates

Reviewer: I am on the fence about including this section at all (and the grafana dashboard config in the docker-compose).  Static config files for Grafana provisioning seems pretty tangentially related to the main topic of custom SQL metrics in Watchdog.  Most of this should just be covered in the Grafana documentation and is not specific to Watchdog.

Personally, I prefer to use version-controlled configuration for my deployments whenever possible, but I don't think that is really how most IT teams actually use Grafana.  Provisioning Grafana with files makes it _very annoying_ to actually modify/improve dashboards and alerts. I would not recommend it as a common practice unless you have a broad group of people maintaining your Grafana instance and need to use file provisioning and a change-control mechanism....  (Or, for some reason, you are planning to deploy a bunch of Watchdog instances with similar configuration.)

Author: fair!  Even though I just justified having an empty JSON file in the docs, I agree.  Lemme rip this stuff out as it's extra steps that detract from the primary goal - agreed!

thanks for the some clarity on what's most important to document here.  tidied things up in 7f30dae
